### PR TITLE
Feature/get lfile owner

### PR DESCRIPTION
**Pull request by andre-merzky**
cleanup of lockfile class, add more tests, documentation.

### CODE REVIEW

File: src/radical/utils/lockfile.py [hunk 1]
```diff
@@ -45,17 +57,45 @@ class Lockfile(object):
 
     # --------------------------------------------------------------------------
     #
-    def __init__(self, fname):
+    def __init__(self, fname, delete=False, *args, *8kwargs):
```
Comment (mtitov) [inline src/radical/utils/lockfile.py:60]: typo next to `kwargs` (-> `**kwargs`)

File: src/radical/utils/lockfile.py [hunk 2]
```diff
@@ -96,59 +144,114 @@ def acquire(self, timeout=None):
...
+        That name can then be retrieved by any other process or thread which
+        attempts to lock the same file.  Otherwise this method returns
+        'unknown'.
+        '''
+
+        if not self._delete:
+            return 'unknown'
+
+        if not os.path.isfile(self._fname):
+            return None
```
Comment (mtitov) [inline src/radical/utils/lockfile.py:178]: should it return `'unknown'` as well?

p.s. and may be re-group?
```python
        output = 'unknown'
        if self._delete and os.path.isfile(self._fname):
            try:
                with open(self._fname, 'r') as fin:
                    output = fin.read()
            except:
                pass
        return output
```

File: src/radical/utils/lockfile.py [hunk 3]
```diff
@@ -96,59 +144,114 @@ def acquire(self, timeout=None):
...
+
+        try:
+            with open(self._fname, 'r') as fin:
+                return(fin.read())
```
Comment (mtitov) [inline src/radical/utils/lockfile.py:182]: may be better move `return` out from `with` statement?

File: tests/unittests/test_lockfile.py [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
 import time
+import errno
 import pytest
 
+
+import multiprocessing as mp
+import radical.utils   as ru
+
+
 import radical.utils as ru
```
Comment (mtitov) [inline tests/unittests/test_lockfile.py:14]: duplication

File: src/radical/utils/lockfile.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/utils/lockfile.py:60]: Oops...  Well, it did shut up the linter :-P

File: src/radical/utils/lockfile.py (hunk 2 above)
Comment (andre-merzky) [reply src/radical/utils/lockfile.py:178]: > should it return `'unknown'` as well?

When the lockfile does not exist anymore, it is not locked by anybody - so `None` is actually correct.

File: src/radical/utils/lockfile.py (hunk 3 above)
Comment (andre-merzky) [reply src/radical/utils/lockfile.py:182]: Hmm, why?  The return will leave the scope, and the `with` resource context will be released.

File: tests/unittests/test_lockfile.py (hunk 4 above)
Comment (andre-merzky) [reply tests/unittests/test_lockfile.py:14]: Oh, thanks

File: src/radical/utils/lockfile.py (hunk 2 above)
Comment (mtitov) [reply src/radical/utils/lockfile.py:185]: should it be checked first then?

File: src/radical/utils/lockfile.py (hunk 3 above)
Comment (mtitov) [reply src/radical/utils/lockfile.py:189]: yeah, maybe python can handle it, just the construction doesn't look good, but ok

File: src/radical/utils/lockfile.py (hunk 2 above)
Comment (andre-merzky) [reply src/radical/utils/lockfile.py:185]: > should it be checked first then?

Fair, I changed the order.

File: src/radical/utils/lockfile.py (hunk 3 above)
Comment (andre-merzky) [reply src/radical/utils/lockfile.py:189]: Yeah, I think that's fine - it's the whole purpose of the `with` construct really, to ensure cleanup on any form of scope exit...

**Comment by andre-merzky** [comment]
I have no idea why the tests are failing - do you?

**Comment by mtitov** [comment]
@andre-merzky I think it might be related to this [L136](https://github.com/radical-cybertools/radical.utils/blob/b8acfd47116ee8d2f0efc50629140640b452be4d/src/radical/utils/lockfile.py#L136), seems that `fcntl.flock` doesn't return anything (== always returns `None`)

btw should we also consider same locking as for IDs -> [link](https://github.com/radical-cybertools/radical.utils/blob/devel/src/radical/utils/ids.py#L262)

**Comment by andre-merzky** [comment]
> @andre-merzky I think it might be related to this [L136](https://github.com/radical-cybertools/radical.utils/blob/b8acfd47116ee8d2f0efc50629140640b452be4d/src/radical/utils/lockfile.py#L136), seems that `fcntl.flock` doesn't return anything (== always returns `None`)
> 
> btw should we also consider same locking as for IDs -> [link](https://github.com/radical-cybertools/radical.utils/blob/devel/src/radical/utils/ids.py#L262)

`lockf` is not thread safe: any thread releasing a `lockf` lock will also release the lock for all other threads holding it.  I don't mind much for IDs where we usually have a single point in code where that is handled, but I do mind for the generic lock file.  On Theta, it's probably a file system specific issue (do you happen to know on what file system type the `$RADICAL_BASE` lived?

`flock` *is* thread safe - but it is not defined by POSIX, but by BSD.  Even NFS supports it nowadays (under linux), but it's not unlikely that there are shared file systems around which don't.  

Either way, I'd rather not fall back to `lockf` in this case.  Hmmm...  Nothing is ever easy...

**Comment by andre-merzky** [comment]
@mtitov : I have to ask you for a new review on this: I switched from `fcntl` to symlinks for locking...  This should be more portable now.  We probably don't want to use this for the IDs, as it will likely not scale well on shared file systems (it hits the metadata servrs a couple of times).

**Review by mtitov — APPROVED** [review]
wow, completely new code :) LGTM!

p.s. sorry, I wasn't clear with my last comment - I meant not to switch but to use both (with `try...except`), but it is not the case for now. And for Theta it was regarding GPFS

**Comment by andre-merzky** [comment]
>  I meant not to switch but to use both (with `try...except`),

Yes, understood - but that would change the semantics of the lock to 'not thread safe', which I wanted to avoid...

**Pull request merged**